src/lib.rs

```diff
@@ -134,49 +90,77 @@ pub trait BitField: Copy + Eq + Not<Output = Self> +
     /// not in the lower N bits of `value`.
-    fn set_range(&mut self, range: Range<u8>, value: Self) -> &mut Self {
-        assert!(range.start < self.length());
-        assert!(range.end <= self.length());
-        assert!(range.start < range.end);
-        assert!(value << (self.length() - (range.end - range.start)) >>
-                (self.length() - (range.end - range.start)) == value,
-                "value too big");
+    fn set_bits(&mut self, range: Range<u8>, value: Self) -> &mut Self;
+}
+
+/// An internal macro used for implementing BitField on the standard integral types.
+macro_rules! bitfield_numeric_impl {
+    ($($t:ty)*) => ($(
+        impl BitField for $t {
+            fn bit_length(&self) -> u8 {
+                ::core::mem::size_of::<Self>() as u8 * 8
+            }
 
-        let bitmask: Self = !(!Self::zero() << (self.length() - range.end) >>
-                              (self.length() - range.end) >>
-                              range.start << range.start);
+            fn get_bit(&self, bit: u8) -> bool {
+                assert!(bit < self.bit_length());
 
-        let bits = *self & bitmask;
+                (*self & (1 << bit)) != 0
+            }
 
-        // set bits
-        *self = bits | (value << range.start);
+            fn get_bits(&self, range: Range<u8>) -> Self {
+                assert!(range.start < self.bit_length());
+                assert!(range.end <= self.bit_length());
+                assert!(range.start < range.end);
 
-        self
-    }
-}
+                // shift away high bits
+                let bits = *self << (self.bit_length() - range.end) >> (self.bit_length() - range.end);
 
-/// An internal macro used for implementing BitField on the standard unsigned integral types.
-macro_rules! bitfield_impl {
-    ($($t:ty)*) => ($(
-        impl BitField for $t {
-            fn zero() -> Self { 0 }
-            fn one() -> Self { 1 }
-            fn length(&self) -> u8 {
-                size_of::<Self>() as u8 * 8
+                // shift away low bits
+                bits >> range.start
+            }
+
+            fn set_bit(&mut self, bit: u8, value: bool) -> &mut Self {
+                assert!(bit < self.bit_length());
+
+                if value {
+                    *self |= 1 << bit;
+                } else {
+                    *self &= !(1 << bit);
+                }
+
+                self
+            }
+
+            fn set_bits(&mut self, range: Range<u8>, value: Self) -> &mut Self {
+                assert!(range.start < self.bit_length());
+                assert!(range.end <= self.bit_length());
+                assert!(range.start < range.end);
+                assert!(value << (self.bit_length() - (range.end - range.start)) >>
+                        (self.bit_length() - (range.end - range.start)) == value,
+                        "The provided value when setting a range of bits had zeros outside of the size of the range!");
```

Comment:
I don't understand what you're trying to say here. I also think it's better to keep this message short and understandable for the common case (e.g. "value too large" or "value does not fit into bit range").